### PR TITLE
Chore/processed blocks table

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
Creates structures in place for the blockchain indexer. Todo:

- [x] db migration script
- [x] env handling to enable indexer
- [x] basic indexer structure

### CODE REVIEW

File: src/lib/services/blockchainWatcher.ts [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
-const typeTableMap = {
-  [TokenType.DEMAND]: 'demand',
-  [TokenType.MATCH2]: 'match2',
+const typeToTable = (tokenType: TOKEN_TYPE) => {
+  switch (tokenType) {
+    case 'DEMAND':
+      return 'demand'
+    case 'MATCH2':
+      return 'match2'
+  }
```
Comment (n3op2) [inline src/lib/services/blockchainWatcher.ts:16]: const typeToTable = (tokenType: TOKEN_TYPE) => tokenType.toLowerCase()? why switch? Just out of interest

**Review by n3op2 — APPROVED** [review]

File: src/lib/db/migrations/20230331084321_processed-blocks-table.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    def.specificType('hash', 'CHAR(64)').notNullable()
+    def.bigInteger('height').unsigned().notNullable().unique()
+    def.specificType('parent', 'CHAR(64)').notNullable()
+    def.datetime('created_at').notNullable().defaultTo(knex.fn.now())
+
+    def.primary(['hash'])
+  })
+
+  await knex.schema.alterTable('processed_blocks', (def) => {
+    def.foreign('parent', 'fk_processed_blocks_parent_hash').references('hash').inTable('processed_blocks')
```
Comment (jonmattgray) [inline src/lib/db/migrations/20230331084321_processed-blocks-table.ts:14]: how would the first insert work? Seems like a chicken/egg thing, you need an existing row with a hash to set as parent, and that row needs a row for its parent and so on...

Comment (jonmattgray) [reply src/lib/db/migrations/20230331084321_processed-blocks-table.ts:14]: since the columns are `NotNullable`

Comment (mattdean-digicatapult) [reply src/lib/db/migrations/20230331084321_processed-blocks-table.ts:14]: A row can reference itself

File: src/lib/services/blockchainWatcher.ts (hunk 1 above)
Comment (mattdean-digicatapult) [reply src/lib/services/blockchainWatcher.ts:16]: it forces the correct type coertion. I tried to do this with an object map but typescript wasn't happy with that

File: src/lib/indexer/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { Logger } from 'pino'
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply src/lib/indexer/index.ts:1]: I hate it to be honest. I'm really not a fan of OOP and class structures but I'm trying to be consistent with the style of the repo

**Pull request merged**